File: agent-wechat-fork/agent-server-rust/src/bridge/mention_names.rs
//! Resolve display names for group @-mentions when sending replies.

const HAIR_SPACE: char = '\u{2005}';

/// Names that cannot be used in the WeChat @ picker search.
pub fn is_mentionable_display_name(name: &str) -> bool {
    let trimmed = name.trim();
    !trimmed.is_empty() && !trimmed.starts_with("wxid_")
}

/// Extract `@Name` tokens terminated by hair space (WeChat mention protocol).
pub fn extract_at_display_names(text: &str) -> Vec<String> {
    let mut names = Vec::new();
    for segment in text.split(HAIR_SPACE) {
        let token = segment.trim();
        if token.starts_with('@') || token.starts_with('\u{FF20}') {
            let name = token.trim_start_matches(['@', '\u{FF20}']).trim();
            if is_mentionable_display_name(name) {
                names.push(name.to_string());
            }
        }
    }
    names
}

/// Pick the best single name to @ when replying to a group message sender.
pub fn resolve_for_reply(
    sender_name: Option<&str>,
    raw_content: Option<&str>,
    mention_display_name: Option<&str>,
) -> Vec<String> {
    if let Some(name) = mention_display_name.filter(|n| is_mentionable_display_name(n)) {
        return vec![name.to_string()];
    }
    if let Some(name) = sender_name.filter(|n| is_mentionable_display_name(n)) {
        return vec![name.to_string()];
    }
    if let Some(content) = raw_content {
        if let Some(name) = extract_at_display_names(content).into_iter().next() {
            return vec![name];
        }
    }
    Vec::new()
}

/// Remove leading `@Name` tokens the LLM echoed when FSM will insert real mentions.
pub fn strip_leading_at_mentions(text: &str, mentions: &[String]) -> String {
    let mut rest = text.trim_start();
    for name in mentions {
        for prefix in [format!("@{name}"), format!("@{name}\u{2005}")] {
            if rest.starts_with(&prefix) {
                rest = rest[prefix.len()..].trim_start();
            }
        }
    }
    rest.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_extract_at_tokens_hair_space() {
        let text = "@Agent\u{2005}hello world";
        assert_eq!(extract_at_display_names(text), vec!["Agent"]);
    }

    #[test]
    fn test_resolve_prefers_mention_display_name() {
        let names = resolve_for_reply(Some("wxid_abc"), Some("@Bot\u{2005}hi"), Some("Leaif"));
        assert_eq!(names, vec!["Leaif"]);
    }

    #[test]
    fn test_resolve_uses_sender_name() {
        let names = resolve_for_reply(Some("Alice"), None, None);
        assert_eq!(names, vec!["Alice"]);
    }

    #[test]
    fn test_rejects_wxid_sender() {
        assert!(resolve_for_reply(Some("wxid_abc"), None, None).is_empty());
    }
}

File: packages/agent/src/mention-names.ts
/** WeChat @-mention protocol uses hair space (U+2005) after display names. */
const HAIR_SPACE = "\u2005";

export function isMentionableDisplayName(name: string): boolean {
  const trimmed = name.trim();
  return trimmed.length > 0 && !trimmed.startsWith("wxid_");
}

export function extractAtDisplayNames(text: string): string[] {
  const names: string[] = [];
  for (const segment of text.split(HAIR_SPACE)) {
    const token = segment.trim();
    if (token.startsWith("@") || token.startsWith("\uFF20")) {
      const name = token.replace(/^[@\uFF20]+/, "").trim();
      if (isMentionableDisplayName(name)) {
        names.push(name);
      }
    }
  }
  return names;
}

export function resolveForReply(
  senderName: string | undefined,
  rawContent: string | undefined,
  mentionDisplayName: string | undefined,
): string[] {
  if (mentionDisplayName && isMentionableDisplayName(mentionDisplayName)) {
    return [mentionDisplayName];
  }
  if (senderName && isMentionableDisplayName(senderName)) {
    return [senderName];
  }
  if (rawContent) {
    const first = extractAtDisplayNames(rawContent)[0];
    if (first) return [first];
  }
  return [];
}

/** Strip leading @Name tokens the model echoed when FSM inserts real mentions. */
export function stripLeadingAtMentions(text: string, mentions: string[]): string {
  let rest = text.trimStart();
  for (const name of mentions) {
    for (const prefix of [`@${name}`, `@${name}${HAIR_SPACE}`]) {
      if (rest.startsWith(prefix)) {
        rest = rest.slice(prefix.length).trimStart();
      }
    }
  }
  return rest;
}
